server: TestReportUsage failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/cee0b362d9f982446444b4a704db86e0cbac7c6d

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReportUsage PKG=github.com/cockroachdb/cockroach/pkg/server TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1339976&tab=buildLog

```
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
W190615 06:12:14.624248 46550 storage/intentresolver/intent_resolver.go:822  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I190615 06:12:14.624270 46030 util/stop/stopper.go:546  quiescing; tasks left:
5      [async] storage.IntentResolver: cleanup txn records
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624297 46030 util/stop/stopper.go:546  quiescing; tasks left:
4      [async] storage.IntentResolver: cleanup txn records
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624323 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
3      [async] storage.IntentResolver: cleanup txn records
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624352 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
2      [async] storage.IntentResolver: cleanup txn records
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624383 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624418 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      node.Node: writing summary
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
W190615 06:12:14.624464 46284 server/node.go:839  [n1,summaries] error recording initial status summaries: failed to send RPC: sending to all 1 replicas failed; last error: (err: [NotLeaseHolderError] r3: replica (n1,s1):1 not lease holder; lease holder unknown) <nil>
I190615 06:12:14.624491 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624538 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
1      [async] intent_resolver_gc_batcher
I190615 06:12:14.624559 46030 util/stop/stopper.go:546  quiescing; tasks left:
3      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
I190615 06:12:14.634404 46030 util/stop/stopper.go:546  quiescing; tasks left:
2      storage.IntentResolver: processing intents
2      [async] closedts-rangefeed-subscriber
I190615 06:12:14.634451 46030 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] closedts-rangefeed-subscriber
1      storage.IntentResolver: processing intents
I190615 06:12:14.634479 46030 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
--- FAIL: TestReportUsage (0.19s)
    updates_test.go:404: error applying zone config "constraints: [+zone=somestring, +somestring]" to "TABLE system.rangelog": pq: constraint "+zone=somestring" matches no existing nodes within the cluster - did you enter it correctly?

```

--- a/pkg/server/updates.go
+++ b/pkg/server/updates.go
@@ -19,6 +19,7 @@ import (
 	"io"
 	"io/ioutil"
 	"math/rand"
+	"net"
 	"net/http"
 	"net/url"
 	"reflect"
@@ -454,7 +455,19 @@ func (s *Server) reportDiagnostics(ctx context.Context) {
 		return
 	}
 	addInfoToURL(ctx, reportingURL, s, report.Node)
-	res, err := http.Post(reportingURL.String(), "application/x-protobuf", bytes.NewReader(b))
+
+	const timeout = 3 * time.Second
+	client := &http.Client{
+		Transport: &http.Transport{
+			// Don't leak a goroutine on OSX (the TCP level timeout is probably
+			// much higher than on linux) when the connection fails in weird ways.
+			DialContext:       (&net.Dialer{Timeout: timeout}).DialContext,
+			DisableKeepAlives: true,
+		},
+		Timeout: timeout,
+	}
+
+	res, err := client.Post(reportingURL.String(), "application/x-protobuf", bytes.NewReader(b))
 	if err != nil {
 		if log.V(2) {
 			// This is probably going to be relatively common in production
@@ -464,9 +477,8 @@ func (s *Server) reportDiagnostics(ctx context.Context) {
 		return
 	}
 	defer res.Body.Close()
-
-	if res.StatusCode != http.StatusOK {
-		b, err := ioutil.ReadAll(res.Body)
+	b, err = ioutil.ReadAll(res.Body)
+	if err != nil || res.StatusCode != http.StatusOK {
 		log.Warningf(ctx, "failed to report node usage metrics: status: %s, body: %s, "+
 			"error: %v", res.Status, b, err)
 	}

--- a/pkg/server/updates_test.go
+++ b/pkg/server/updates_test.go
@@ -225,6 +225,7 @@ func TestCBOReportUsage(t *testing.T) {
 	s, db, _ := serverutils.StartServer(t, params)
 	// Stopper will wait for the update/report loop to finish too.
 	defer s.Stopper().Stop(context.TODO())
+	defer db.Close()
 	ts := s.(*TestServer)
 
 	// make sure the test's generated activity is the only activity we measure.
@@ -396,11 +397,15 @@ func TestReportUsage(t *testing.T) {
 		{"DATABASE system", fmt.Sprintf(`constraints: {"+zone=%[1]s,+%[1]s": 2, +%[1]s: 1}`, elemName)},
 		{"DATABASE system", fmt.Sprintf(`experimental_lease_preferences: [[+zone=%[1]s,+%[1]s], [+%[1]s]]`, elemName)},
 	} {
-		if _, err := db.Exec(
-			fmt.Sprintf(`ALTER %s CONFIGURE ZONE = '%s'`, cmd.resource, cmd.config),
-		); err != nil {
-			t.Fatalf("error applying zone config %q to %q: %v", cmd.config, cmd.resource, err)
-		}
+		testutils.SucceedsSoon(t, func() error {
+			if _, err := db.Exec(
+				fmt.Sprintf(`ALTER %s CONFIGURE ZONE = '%s'`, cmd.resource, cmd.config),
+			); err != nil {
+				// Work around gossip asynchronicity.
+				return errors.Errorf("error applying zone config %q to %q: %v", cmd.config, cmd.resource, err)
+			}
+			return nil
+		})
 	}
 	if _, err := db.Exec(`INSERT INTO system.zones (id, config) VALUES (10000, null)`); err != nil {
 		t.Fatal(err)

--- a/pkg/util/cloudinfo/cloudinfo.go
+++ b/pkg/util/cloudinfo/cloudinfo.go
@@ -15,6 +15,7 @@ package cloudinfo
 import (
 	"encoding/json"
 	"io/ioutil"
+	"net"
 	"net/http"
 	"regexp"
 	"time"
@@ -91,8 +92,15 @@ type metadataReqHeader struct {
 }
 
 func getInstanceMetadata(url string, headers []metadataReqHeader) ([]byte, error) {
+	const timeout = 500 * time.Millisecond
 	client := http.Client{
-		Timeout: 500 * time.Millisecond,
+		Timeout: timeout,
+		Transport: &http.Transport{
+			// Don't leak a goroutine on OSX (the TCP level timeout is probably
+			// much higher than on linux).
+			DialContext:       (&net.Dialer{Timeout: timeout}).DialContext,
+			DisableKeepAlives: true,
+		},
 	}
 
 	req, err := http.NewRequest("GET", url, nil)

